File: frontend/apps/app/android/app/src/main/kotlin/live/pyxietarot/app/AuthBridgePlugin.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package live.pyxietarot.app

import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

private const val WIDGET_PREFS_NAME = "widget_prefs"
private const val AUTH_TOKEN_KEY = "auth_token"

/** Mirrors the web app's JWT (localStorage-only otherwise) into native storage, so the widget's
 * background worker can authenticate without the WebView running; also exposes an explicit refresh
 * trigger for app-side events (e.g. a new diary entry) that don't otherwise touch the token. */
@CapacitorPlugin(name = "AuthBridge")
class AuthBridgePlugin : Plugin() {
    private fun prefs() = context.getSharedPreferences(WIDGET_PREFS_NAME, 0)

    @PluginMethod
    fun setToken(call: PluginCall) {
        val token = call.getString("token") ?: return call.reject("token is required")
        prefs().edit().putString(AUTH_TOKEN_KEY, token).apply()
        SpreadWidgetScheduler.refreshNow(context)
        call.resolve()
    }

    @PluginMethod
    fun clearToken(call: PluginCall) {
        prefs().edit().remove(AUTH_TOKEN_KEY).apply()
        SpreadWidgetScheduler.refreshNow(context)
        call.resolve()
    }

    @PluginMethod
    fun refreshWidget(call: PluginCall) {
        SpreadWidgetScheduler.refreshNow(context)
        call.resolve()
    }
}

File: frontend/apps/app/ios/App/App/AuthBridgePlugin.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
import Capacitor
import WidgetKit

/// Mirrors the web app's access token into the App Group so the widget can authenticate without the WebView
/// running; also exposes an explicit refresh trigger for app-side events (e.g. a new diary entry).
@objc(AuthBridgePlugin)
class AuthBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    let identifier = "AuthBridgePlugin"
    let jsName = "AuthBridge"
    let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refreshWidget", returnType: CAPPluginReturnPromise),
    ]

    @objc func setToken(_ call: CAPPluginCall) {
        WidgetStore.token = call.getString("token")
        refreshWidget(call)
    }

    @objc func clearToken(_ call: CAPPluginCall) {
        WidgetStore.token = nil
        refreshWidget(call)
    }

    @objc func refreshWidget(_ call: CAPPluginCall) {
        WidgetCenter.shared.reloadTimelines(ofKind: WidgetStore.kind)
        call.resolve()
    }
}
